#                                        Assignment 2

In [1]:
#IMPORTING REQUIRED MODULES

import numpy as np
import pandas as pd
import warnings
import math
import time
warnings.filterwarnings('ignore')

In [2]:
#GRADIENT DESCENT FUNCTION GIVEN IN ASSIGNMENT

def gradient_descent(gradient,init_,learn_rate,n_iter=100,tol=1e-06):
    x=init_
    for i in range(n_iter):
        delta= -learn_rate * gradient(x)
        if np.all(np.abs(delta)<=tol):
            break
        x+=delta
    return round(x*1000)/1000


# 1.a(i) finding minima for $f(x) = x^2+3x+4$

In [3]:
#VARIABLES TO STORE THE MINIMA AND VALUE OF X AT WHICH THE MINIMA OCCURS

min_x=0
minima=100000

#CALCULATING THE GRADIENT DESCENT FOR DIFFERENT LEARNING RATES(i*0.001 is the learning rate for that iteration)

for i in range(0,50):
    
    #GRADIENT FOR X^2+3X+4 IS 2X+3(THIS IS DEFINED AS GRADIENT FUNCTION)
    #CALLING GRADIENT DESCENT FUNCTION
    #STORING THE X VALUE RETURNED BY GRADIENT DESCENT IN current_x 
    
    current_x=gradient_descent(gradient=lambda v: 2 * v + 3, init_=4, learn_rate=0.001*i)
    
    #CALCULATING THE VALUE OF X^2+3X+4 FOR THAT current_x VALUE AND STORING IT IN current_minima
    
    current_minima=(current_x)**2+3*current_x+4
    
    #IF CURRENT MINIMA IS LESS THAN OVERALL MINIMA CALCULATED TILL THAT ITERATION THEN UPDATING
    #MINIMA AS CURRENT MINIMA AND UPDATING THE MIN_X VALUE VALUE TO X VALUE CORRESPONDING TO THAT CURRENT MINIMA
    
    if current_minima<=minima:
        min_x=current_x
        minima=current_minima

#PRINTING THE OVERALL MINIMA OF X^2+3X+4 AND THE VALUE X WHERE THE MINIMA OCCURS

print('minima of function x^2 + 3x +4 is:'+str(minima)+' at x value:'+str(min_x))

minima of function x^2 + 3x +4 is:1.75 at x value:-1.5


# 1.a(ii) finding minima for $f(x) = x^4-3x^2+2x$

In [4]:
#VARIABLES TO STORE THE MINIMA AND VALUE OF X AT WHICH THE MINIMA OCCURS

min_x=0
minima=100000

#CALCULATING THE GRADIENT DESCENT FOR DIFFERENT LEARNING RATES(i*0.001 is the learning rate for that iteration)

for i in range(1,30):
    
    #GRADIENT FOR X^4-3X^2+2X IS 4X^3-6X+2(THIS IS DEFINED AS GRADIENT FUNCTION)
    #CALLING GRADIENT DESCENT FUNCTION
    #STORING THE X VALUE RETURNED BY GRADIENT DESCENT IN current_x 
    
    current_x=gradient_descent(gradient=lambda v:  4* v*v*v-6*v + 2, init_=4.0, learn_rate=0.001*i)
    
     #CALCULATING THE VALUE OF X^4-3X^2+2X FOR THAT current_x VALUE AND STORING IT IN current_minima
    
    current_minima=(current_x)**4-3*(current_x)**2+2*current_x
    
    #IF CURRENT MINIMA IS LESS THAN OVERALL MINIMA CALCULATED TILL THAT ITERATION THEN UPDATING
    #MINIMA AS CURRENT MINIMA AND UPDATING THE MIN_X VALUE VALUE TO X VALUE CORRESPONDING TO THAT CURRENT MINIMA
    
    if current_minima<=minima:
        min_x=current_x
        minima=current_minima
        
#PRINTING THE OVERALL MINIMA OF X^4-3X^2+2X  AND THE VALUE X WHERE THE MINIMA OCCURS

print('minima of function x^4-3x^2+2x is:'+str(minima)+' at x value:'+str(min_x))

minima of function x^4-3x^2+2x is:-4.848076206064 at x value:-1.366


# 1.b gradient function to calculate gradients for a linear regression y = ax+b
# Loss Function $L(a,b) = \frac{1}{2n} \sum_{i=1}^n (y_{i}-ax_{i}-b)^2$
# $\frac{\partial (L)}{\partial a} = \frac{1}{n} \sum_{i=1}^n (y_{i}-ax_{i}-b)(-x_{i})$
# $\frac{\partial (L)}{\partial b} = \frac{1}{n} \sum_{i=1}^n -(y_{i}-ax_{i}-b)$

In [5]:
#GRADIENT FUNCTION TO CALCULATE GRADIENTS FOR LINEAR REGRESSION(y=ax+b)

#FUNCTION TAKES PARAMETERS W(WEIGHTS VECTOR),X(INPUT FEATURES MATRIX),Y(OUTPUT FEATURE MATRIX)

def gradient(w,X,Y):
    
    #CALCULATING THE PREDICTED OUTPUT MATRIX BASED ON CUURENT WEIGHT VECTOR PASSED TO FUNCTION
    
    y_pred = X.dot(w)
    
    #CALCULATING THE GRADIENT FOR LINEAR REGRESSION
    #LOSS FUNCTION IS 1/2(Y-y_pred)^2
    
    grad_w = -(Y - y_pred).dot(X)
    
    #RETURNING THE AVERAGE GRADIENT VECTOR
    
    return grad_w/len(X)

In [6]:
# GRADIENT DESCENT FUNCTION FOR LINEAR REGRESSION

def gradient_descent(gradient,init_,learn_rate,n_iter=100,tol=1e-06):
    
    #INITIALIZING THE WEIGHT VECTOR
    
    w=init_
    
    #UNTIL THE NO OF ITERATIONS 
    
    for i in range(n_iter):
        
        #GRADIENT FUNCTION CALCULATES GRADIENT FOR LINEAR REGRESSION
        
        delta= -learn_rate * gradient(w,X,Y)
        
        #IF DELTA IS <= TOL THEN BREAK THE LOOP(SINCE THE UPDATES ARE VERY SMALL i.e CONVERGED)
        
        if np.all(np.abs(delta)<=tol):
            break
        
        #WEIGHT VECTOR UPDATING
        
        w+=delta
    
    #RETURNING WEIGHT VECTOR
    
    return w

# 1.c using gradient descent to find the optimal parameters relating X with Y

In [7]:
#GENERATING ARTIFICIAL DATA FOR REGRESSION

np.random.seed(0)
X=2.5 * np.random.randn(1000000)+1.5
res=1.5* np.random.randn(1000000)
Y=2+ 0.3*X + res

#INSERTING CONSTANT ONES FOR BIAS WEIGHTS TO INPUT FEATURE MATRIX

X= X[..., np.newaxis]
X = np.insert(X, 0, 1, axis=1)

In [8]:
#CREATING VARIABLES TO STORE MINIMUM MSE AND WEIGHTS CORRESPONDING TO THAT MINIMUM MSE

minimum_mse=100000
opt_w=[0]
optimal_lr=0

#STROING THE TIME AT THIS POINT OF EXECTION TO CALCULATE TIME TAKEN TO RUN THE BELOW CODE

start_time = time.time()

#ITERATING THROUGH DIFFERENT LEARNING RATES

for i in [0.2,0.02,0.002]:
    
    #CALLING GRADIENT DESCENT FUNCTION FOR LINEAR REGRESSION WITH THAT ITERATIONS LEARNING RATE
    #AND STORING THE WEIGHT VECTOR RETURNED IN current_w
    
    current_w=gradient_descent(gradient, init_=[0,0], learn_rate=i,n_iter=300)
    
    #CALCULATING THE CURRENT PREDICTED Y VALUES BASED ON THE current_w
    
    current_y_pred = X.dot(current_w)
    
    #CALCULATING CURRENT MSE BASED ON CURRENT PREDICTED Y VALUES
    
    current_mse = np.mean(0.5*(Y - current_y_pred)**2 )
    
    #IF CURRENT MSE IS LESS THAN MINIMUM MSE AND CURRENT MSE IS NOT NA
    #WE UPDATE THE MINIMUM MSE AS CURRENT MSE AND WEIGHTS CORRESPONDING TO MINIMUM MSE AS OPTIMAL WEIGHTS AND
    #THE LEARNING RATE CORRESPONDING TO CURRENT MSE AS OPTIMAL LEARNING RATE
    
    if current_mse!=math.nan and current_mse<=minimum_mse:
        minimum_mse=current_mse
        opt_w[0]=current_w
        optimal_lr=i
        
#PRINTING THE OPTIMAL WEIGHTS VECTOR i.e(a,b) AND OPTIMAL LEARNING RATE CORRESPONDING TO MIN MSE

print('optimal parameters relating X with Y are: a:'+str(opt_w[0][1])+'  b:'+str(opt_w[0][0])) 
print('we get these optimal weights at learning rate:'+str(optimal_lr)+' and mse corresponding to these weights is:'+str(minimum_mse))

#STORING THE TIME TAKEN TO RUN GRADIENT DESCENT WITH 300 EPOCHES IN gd_time

gd_time=time.time() - start_time

#PRINTING THE TIME TAKEN TO RUN GRADIENT DESCENT FOR LINEAR REGRESSION

print("--- %s seconds ---" % (gd_time))

optimal parameters relating X with Y are: a:0.30037444363643423  b:1.999888927889675
we get these optimal weights at learning rate:0.2 and mse corresponding to these weights is:1.1224947962169465
--- 12.645088195800781 seconds ---


# 1.d  Implementing minibatch stochastic gradient descent

In [9]:
#GRADIENT FUCTION FOR MINIBATCH SGD(SAME AS GRADIENT FUNCTION FOR GRADIENT DESCENT)

def gradient(w,X,Y):
    
    #CALCULATING THE PREDICTED OUTPUT MATRIX BASED ON CUURENT WEIGHT VECTOR PASSED TO FUNCTION
    
    y_pred = X.dot(w)
    
    #CALCULATING THE GRADIENT FOR LINEAR REGRESSION
    #LOSS FUNCTION IS 1/2(Y-y_pred)^2
    
    grad_w = -(Y - y_pred).dot(X)
    
    #RETURNING THE AVERAGE GRADIENT VECTOR
    
    return grad_w/len(X)

In [10]:
#MINI BATCH SGD FUNCTION FOR MINI BATCH SGD

def minibatch_sgd(gradient,mini_batches,init_,learn_rate,tol=1e-06,epoches=50):
    
    #INITIALIZING THE WEIGHT VECTOR
    
    wgt=init_
    
    #CALCULATING TOTAL NO OF MINI BATCHES
    
    total_batches=len(mini_batches)
    
    #STORING CURRENT BATCH NUMBER AND EPOCH NUMBER IN VARIABLES
    
    batch_no=0
    epoch_no=0
    
    
    while(True):
        
        #MINI BATCHES IS A LIST CONTAINING ALL THE BATCHES
        
        #GETS ONLY THE BATCH CORRESPONDING TO THAT batch_no FROM MINI BATCHES LIST
        
        batch=mini_batches[batch_no]
        
        #batch[0],batch[1] CONTAINS INPUT FEATURE MATRIX AND OUTPUT MATRIX CORRESPONDING TO THAT MINI BATCH
        
        #GRADIENT FUNCTION RETURN THE GRADIENT CORRESPONDING TO THAT MINI BATCH
        
        delta= -learn_rate * gradient(wgt,batch[0],batch[1])
        
        #IF DELTA IS <= TOL THEN BREAK THE LOOP(SINCE THE UPDATES ARE VERY SMALL I.E CONVERGED)
        
        if np.all(np.abs(delta)<=tol) :
            break
            
        #WEIGHT VECTOR UPDATING
        
        wgt+=delta
        
        #IF ANY OF THE WEIGHTS IS NA THEN BREAK THE LOOP
        
        if pd.isna(float(wgt[0])) or pd.isna(float(wgt[1])):
            break
        
        #INCREASING THE BATCH NO BY 1 (SO THAT IN NEXT ITERATION GRADIENT IS CALCULATED FOR NEXT BATCH)
        
        batch_no=(batch_no+1)%total_batches
        
        #IF WE REACH THE BATCH NO '0' AGAIN THEN IT MEANS 1 EPOCH IS COMPLETED
        #SO INCREASING THE EPOCH BY 1 IF BATCH NO 'O' IS ITERATED AGAIN
        
        if batch_no==0:
            epoch_no+=1
            
        #IF THE NO OF EPOCHES DONE ARE 10 THEN BREAK THE LOOP
        
        if epoch_no==epoches:
            break
            
    #RETURNING WEIGHT VECTOR
    
    return wgt

In [11]:
#FUNCTION TO GENERATE BATCHES WHEN GIVEN INPUT AND OUTPUT MATRICES AND BATCH SIZE

def batch_generator(X, Y, batch_size):

    #FINDING THE TOTAL NO OF DATA POINTS
    
    n_samples = X.shape[0]
    
    #LOOPING THROUGH TOTAL NO OF BATCHES
    
    for i in np.arange(0, n_samples, batch_size):
        
        #FINDING BEGINING AND ENDING INDICES OF THAT BATCH
        
        begin, end = i, min(i+batch_size, n_samples)
        
        #RETURNING THAT BATCH IN A GENERATOR VARIABLE
        
        yield X[begin:end], Y[begin:end]

In [12]:
#CREATING VARIABLES TO STORE MINIMUM MSE AND WEIGHTS CORRESPONDING TO THAT MINIMUM MSE AND LEARNING RATE CORRESPONDING TO 
#THAT MINIMUM MSE

minimum_mse=100000
optimal_lr=0
optimal_wgts=[0,0]


#SHUFFLING THE INPUT AND OUPUT MATRICES AND STORING THEM IN NEW VARIABLES

np.random.seed(0)
idx = np.arange(X.shape[0])
np.random.shuffle(idx)
X_new=X[idx]
Y_new=Y[idx]

#CREATING A LIST TO STORE MINI BATCHES

mini_batches=[]

#GIVING BATCH SIZE AS 1000

batch_size=1000

#CALLING BATCH GENERATOR FUNCTION WITH SHUFFELED MATRICES AND BATCH SIZE AS 1000
#AND APPENDING THE MINI BATCHES RETURNED BY BATCH GENERATOR FUNCTION INTO mini batches LIST

for mini_batch in batch_generator(X_new,Y_new,batch_size):
    mini_batches.append(mini_batch)

#STROING THE TIME AT THIS POINT OF EXECTION TO CALCULATE TIME TAKEN TO RUN THE MINI BATCH SGD

start_time = time.time()

#ITERATING THROUGH DIFFERENT LEARNING RATES

for learning_rate in [0.2,0.02,0.002]:
       
        #CALLING GRADIENT DESCENT CORRESPONDING TO MINI BATCH SGD WITH EPOCHES SET TO 300
        #STORING THE WEIGHT VECTOR RETURNED BY GRADIENT DESCENT FUNCTION IN current_w
        
        current_w=minibatch_sgd(gradient,mini_batches, init_=[0,0], learn_rate=learning_rate,epoches=300)
        
        #CALCULATING THE CURRENT PREDICTED Y VALUES BASED ON THE current_w
        
        current_y_pred = X.dot(current_w)
        
        #CALCULATING THE CURRENT MSE BASED ON CURRENT PREDICTED Y VALUES
        
        current_mse = np.mean(0.5 * (Y - current_y_pred)**2 )
        
        
        #IF CURRENT MSE IS LESS THAN MINIMUM MSE AND CURRENT MSE IS NOT NA
        #WE UPDATE THE MINIMUM MSE AS CURRENT MSE AND WEIGHTS CORRESPONDING TO MINIMUM MSE AS OPTIMAL WEIGHTS
        #AND THE LEARNING RATE CORRESPONDING TO CURRENT MSE AS OPTIMAL LEARNING RATE
        
        if current_mse!=math.nan and current_mse<=minimum_mse:
            minimum_mse=current_mse
            optimal_wgts=current_w
            optimal_lr=learning_rate

#PRINTING THE OPTIMAL WEIGHTS VECTOR i.e(a,b) AND LEARNING RATE CORRESPONDING TO MIN MSE

print('optimal parameters relating X with Y using mini batch sgd with batch size'+str(batch_size)+' are: a:'+str(optimal_wgts[1])+'  b:'+str(optimal_wgts[0])) 
print('we get these optimal weights at learning rate:'+str(optimal_lr)+' and mse corresponding to these weights is:'+str(minimum_mse))



#STORING THE TIME TAKEN TO RUN MINIBATCH SGD WITH 300 EPOCHES IN minisgd_time

minisgd_time=time.time() - start_time

#PRINTING THE TIME TAKEN TO RUN MINI BATCH SGD WITH BATCH SIZE 1000 FOR LINEAR REGRESSION

print("--- %s seconds ---" % (minisgd_time))

optimal parameters relating X with Y using mini batch sgd with batch size1000 are: a:0.29828298958956334  b:1.9993468066618756
we get these optimal weights at learning rate:0.002 and mse corresponding to these weights is:1.1225152617833036
--- 47.87827157974243 seconds ---


# 1.e implementing sgd(mini batch sgd with batch size=1)

In [13]:
#CREATING VARIABLES TO STORE MINIMUM MSE AND WEIGHTS CORRESPONDING TO THAT MINIMUM MSE AND LEARNING RATE CORRESPONDING TO 
#THAT MINIMUM MSE

minimum_mse=100000
optimal_lr=0
optimal_wgts=[0,0]


#SHUFFLING THE INPUT AND OUPUT MATRICES AND STORING THEM IN NEW VARIABLES

np.random.seed(0)
idx = np.arange(X.shape[0])
np.random.shuffle(idx)
X_new=X[idx]
Y_new=Y[idx]

#CREATING A LIST TO STORE MINI BATCHES

mini_batches=[]

#GIVING BATCH SIZE AS 1(SGD)

batch_size=1

#CALLING BATCH GENERATOR FUNCTION WITH SHUFFELED MATRICES AND BATCH SIZE AS 1000
#AND APPENDING THE MINI BATCHES RETURNED BY BATCH GENERATOR FUNCTION INTO mini batches LIST

for mini_batch in batch_generator(X_new,Y_new,batch_size):
    mini_batches.append(mini_batch)

#STROING THE TIME AT THIS POINT OF EXECTION TO CALCULATE TIME TAKEN TO RUN THE MINI BATCH SGD

start_time = time.time()

#ITERATING THROUGH DIFFERENT LEARNING RATES

for learning_rate in [0.2,0.02,0.002]:
       
        #CALLING GRADIENT DESCENT CORRESPONDING TO MINI BATCH SGD WITH EPOCHES SET TO 300
        #STORING THE WEIGHT VECTOR RETURNED BY GRADIENT DESCENT FUNCTION IN current_w
        
        current_w=minibatch_sgd(gradient,mini_batches, init_=[100,100], learn_rate=learning_rate,epoches=300)
        
        #CALCULATING THE CURRENT PREDICTED Y VALUES BASED ON THE current_w
        
        current_y_pred = X.dot(current_w)
        
        #CALCULATING THE CURRENT MSE BASED ON CURRENT PREDICTED Y VALUES
        
        current_mse = np.mean(0.5 * (Y - current_y_pred)**2 )
        
        
        #IF CURRENT MSE IS LESS THAN MINIMUM MSE AND CURRENT MSE IS NOT NA
        #WE UPDATE THE MINIMUM MSE AS CURRENT MSE AND WEIGHTS CORRESPONDING TO MINIMUM MSE AS OPTIMAL WEIGHTS
        #AND THE LEARNING RATE CORRESPONDING TO CURRENT MSE AS OPTIMAL LEARNING RATE
        
        if current_mse!=math.nan and current_mse<=minimum_mse:
            minimum_mse=current_mse
            optimal_wgts=current_w
            optimal_lr=learning_rate

#PRINTING THE OPTIMAL WEIGHTS VECTOR i.e(a,b) AND LEARNING RATE CORRESPONDING TO MIN MSE

print('optimal parameters relating X with Y using mini batch sgd with batch size '+str(batch_size)+' are: a:'+str(optimal_wgts[1])+'  b:'+str(optimal_wgts[0])) 
print('we get these optimal weights at learning rate:'+str(optimal_lr)+' and mse corresponding to these weights is:'+str(minimum_mse))

#STORING TIME TAKEN TO RUN SGD IN sgd_time

sgd_time=time.time() - start_time

#PRINTING THE TIME TAKEN TO RUN MINI BATCH SGD WITH BATCH SIZE 1 FOR LINEAR REGRESSION

print("--- %s seconds ---" % (sgd_time))

optimal parameters relating X with Y using mini batch sgd with batch size 1 are: a:0.29214498013736956  b:2.0107733074843273
we get these optimal weights at learning rate:0.002 and mse corresponding to these weights is:1.1227074538878257
--- 2.279576301574707 seconds ---


In [14]:
#PRINTING THE TIME DIFFERENCE BETWEEN RUNNING GRADIENT DESCENT AND SGD AND MINI BATCH SGD WITH BATCH SIZE 1000 AND SGD

print(gd_time-sgd_time , minisgd_time-sgd_time)



10.365511894226074 45.598695278167725


As you can see that time taken by sgd with 300 epoches is less than gd with 300 epoches.
and also time taken by sgd with 300 epoches is less than mini batch sgd with batch size 1000 and 300 epoches

SO SGD DOES BETTER IN TERMS OF TIME ON OUR DATA

# 1.e finding optimal batch size that works best(HYPERPARAMETER TUNING)

In [15]:
#FUNCTION TO SHUFFLE DATA

def shuffle_data(X, Y):
    np.random.seed(0)
    idx = np.arange(X.shape[0])
    
    #SHUFFLING THE DATA
    
    np.random.shuffle(idx)
    
    #RETURNING SHUFFLED DATA
    
    return X[idx], Y[idx]

In [16]:
#FUNCTION TO SPLIT OUR DATA INTO TRAIN AND TEST SETS

def train_test_split(X, Y, test_size=0.3):
    
    #CALLING SHUFFLE FUNCTION TO SHUFFLE THE DATA BEFORE SPLITTING
    
    X, Y = shuffle_data(X, Y)
    
    #SPLITTING THE DATA INTO TRAIN AND TEST IN THE RATIO SPECIFIED
    
    split_i = len(Y) - int(len(Y) // (1 / test_size))
    X_train, X_test = X[:split_i], X[split_i:]
    Y_train, Y_test = Y[:split_i], Y[split_i:]

    #RETURNING THE SPLIT DATA
    
    return X_train, X_test, Y_train, Y_test

In [17]:
#SPLITTING THE DATA INTO TEST AND TRAIN BY CALLING train_test_split FUNCTION WITH TEST SIZE RATIO AS 0.3 
    
X_train, X_test, Y_train, Y_test=train_test_split(X,Y,test_size=0.3)

In [18]:
#CREATING VARIABLES TO STORE MINIMUM MSE ON TEST SET AND OPTIMAL BATCH SIZE,WEIGHTS ,LEARNING RATE 
#CORRESPONDING TO THAT MINIMUM MSE 

optimal_lr=0
optimal_wgts=[0,0]
minimum_mse=100000000
optimal_batchsize=0

#LOOPING THROUGH DIFFERENT BATCH SIZES

for batch_size in range(1,1000,50):
    
    #SHUFFLING THE TRAIN DATA
    
    X_new, Y_new = shuffle_data(X_train, Y_train)
    
    #CREATING A LIST TO STORE MINI BATCHES
    
    mini_batches=[]
    
    #CALLING BATCH GENERATOR FUNCTION WITH SHUFFELED MATRICES AND BATCH SIZE AS 1000
    #AND APPENDING THE MINI BATCHES RETURNED BY BATCH GENERATOR FUNCTION INTO mini batches LIST
    
    for mini_batch in batch_generator(X_new,Y_new,batch_size):
        mini_batches.append(mini_batch)
    
    #LOOPING THROUGH DIFFERENT LEARNING RATES
    
    for learning_rate in [0.2,0.02,0.002,0.0002,0.00002]:
        
        #CALLING GRADIENT DESCENT CORRESPONDING TO MINI BATCH SGD WITH EPOCHES SET TO 10
        #STORING THE WEIGHT VECTOR RETURNED BY GRADIENT DESCENT FUNCTION IN current_w
        
        current_w=minibatch_sgd(gradient,mini_batches, init_=[100,100], learn_rate=learning_rate,epoches=10)
        
        #CALCULATING THE CURRENT PREDICTED Y VALUES OF TEST DATA BASED ON THE current_w
        
        current_y_pred = X_test.dot(current_w)
        
        #CALCULATING THE CURRENT MSE BASED ON CURRENT PREDICTED Y VALUES OF TEST DATA
        
        current_mse = np.mean(0.5 * (Y_test - current_y_pred)**2 )
        
        #IF CURRENT MSE IS LESS THAN MINIMUM MSE AND CURRENT MSE IS NOT NA
        #WE UPDATE THE MINIMUM MSE AS CURRENT MSE AND WEIGHTS CORRESPONDING TO MINIMUM MSE AS OPTIMAL WEIGHTS
        #AND THE LEARNING RATE CORRESPONDING TO CURRENT MSE AS OPTIMAL LEARNING RATE
        #AND THE BATCH SIZE CORRESPONDING TO CUURENT MSE AS OPTIMAL BATCH SIZE
        
        if current_mse!=math.nan and current_mse<=minimum_mse:
            minimum_mse=current_mse
            optimal_wgts=current_w
            optimal_lr=learning_rate
            optimal_batchsize=batch_size

#PRINTING OPTIMAL BATCH SIZE AND LEARNING RATE TUNED TO THAT OPTIMAL BATCH SIZE AND WEIGHT VECTOR CORRESPONDING TO IT
#AND ITS MSE ON TEST DATA

print('optimal batch size is:'+str(optimal_batchsize)+
       ' learning rate:'+str(optimal_lr)+
      ' weight vector:'+str(optimal_wgts)+
      ' mse on test data: '+str(minimum_mse))

optimal batch size is:401 learning rate:0.002 weight vector:[1.99911664 0.29972309] mse on test data: 1.1260119886838889


1.as you can see the optimal batch size is :401

2.but this optimal batch size is not a constant ,if our data is shuffled differently before making mini batches then you may
get different optimal batch size

3.as the mini batch size gets higher the generalization error also gets higher as it over fits the data




# 2(i)probability that someone has both cold and a fever

## $ P(cold \wedge fever) = P( fever | cold ) P(cold)  $
## $P( fever | cold ) = 0.307$  -   from the table
## $P(cold) = 0.02$            - from the table
## $ P(cold \wedge fever) = (0.307)*(0.02)  $
## $ P(cold \wedge fever) =  0.00614 $

# 2(ii) probability that someone who has a cough has a cold

### $P(cold|cough) = \frac{P(cough \wedge cold)}{P(cough)}$

## To calculate $P(cough \wedge cold)$ we require $P(lung disease)$ first ,so let us calculate it
### $P(lung disease) = P(lung disease|smokes)P(smokes) + P(lung disease|\overline{smokes})P(\overline{smokes})$
#### all the above values can be directly found in the Bayesian network ,so taking those values and substituting
### $P(lung disease) = 0.1009*0.2 + 0.001*0.8$
### $P(lung disease) = 0.02018 + 0.0008$
### $P(lung disease) = 0.02098$
### we get $P(lung disease) = 0.02098$
## now let us calculate  $P(cough \wedge cold)$
### $P(cough \wedge cold) = P(cough \wedge cold \wedge lungdisease) + P(cough \wedge cold \wedge \overline{lungdisease})$
### $=P(cough|cold,lungdisease)P(cold)P(lungdisease)+P(cough|cold,\overline{lungdisease})P(cold)P(\overline{lu  ngdisease})$
#### except $P(lung disease)$ all other values can be directly found in Bayesian network and we have already calculated $P(lung disease)$

### $=0.7525*0.02*0.02098+0.505*0.02*(1-0.02098)$
### $=0.010203851$
### we get $P(cough \wedge cold) =0.010203851$
## now let us calculate  $P(cough)$
### $P(cough) = P(cough|cold,lungdisease)P(cold)P(lungdisease)+P(cough|\overline{cold},lungdisease)P(\overline{cold})P(lungdisease)+P(cough|cold,\overline{lungdisease})P(cold)P(\overline{lungdisease})+P(cough|\overline{cold},\overline{lungdisease})P(\overline{cold})P(\overline{lungdisease})$
#### all the above values can be directly found in Bayesian network except $P(lung disease)$ ,by substituting them we get
### $P(cough) = 0.7525*0.02*0.02098 + 0.505*0.98*0.02098 + 0.505*0.02*0.97902 + 0.01*0.98*0.97902$
### $P(cough) = 0.000315749 + 0.010383002 + 0.009888102 + 0.009594396$
### $P(cough) = 0.030181249$
### we get  $P(cough) = 0.030181249$
## now we have both $P(cough \wedge cold) =0.010203851$ and $P(cough) = 0.030181249$ 
### $P(cold|cough) = \frac{P(cough \wedge cold)}{P(cough)}$
### $P(cold|cough) = \frac{0.010203851}{0.030181249}$
### $P(cold|cough) = 0.338085776$
## finally we get $P(cold|cough) = 0.338085776$

# 3.MLE for the parameters of a k-sided multinomial distribution

### Let The probabilities of the observations are as follows:
### $p_{1}= P(X_{1}) = \frac{x_{1}}{n}$
### $p_{2}= P(X_{2}) = \frac{x_{2}}{n}$
### ....
### $p_{k}= P(X_{k}) = \frac{x_{k}}{n}$
### The likelihood of k sided multinomial distribution as a joint probability $p$ is
### $L(p) = \frac{n!}{x_{1}!x_{2}!..x_{k}!}\prod_{i=1}^{k} p_{i}^{x_{i}}$
### $L(p) = n!\prod_{i=1}^{k} \frac{p_{i}^{x_{i}}}{x_{i}!}$
### Applying log likelihood for the above
### $\log L(p) = \log \left(n!\prod_{i=1}^{k} \frac{p_{i}^{x_{i}}}{x_{i}!}\right)$
###                $\hspace{2.15cm}= \log n! + \log \prod_{i=1}^{k}\frac{p_{i}^{x_{i}}}{x_{i}!}$
###                $\hspace{2.15cm}= \log n! + \sum_{i=1}^k \log\frac{p_{i}^{x_{i}}}{x_{i}!}$
### $\hspace{2.15cm}=\log n! + \sum_{i=1}^k x_i \log p_i - \sum_{i=1}^k \log x_{i}!$
### Let  $\hspace{0.25cm}l(p) = \log L(p)$
### posing a constraint ( $\sum_{i=1}^k {p_{i}}=1$) with lagrange multiplier
### $l^\prime(p,\lambda) = l(p) + \lambda \left(1-\sum_{i=1}^{k} p_{i}\right)$
### $l^\prime(p,\lambda) =\log n! + \sum_{i=1}^k x_i \log p_i - \sum_{i=1}^k \log x_{i}!+\lambda \left(1-\sum_{i=1}^{k} p_{i}\right)$
### Now to find argmax for $l^\prime(p,\lambda)$, lets differentitate w.r.t to $p$
### $\frac{\partial {l^\prime(p,\lambda)}}{\partial p_{i}} = \frac{\partial {}}{\partial p_{i}}\log n!+\frac{\partial {}}{\partial p_{i}}\sum_{i=1}^k x_i \log p_i-\frac{\partial {}}{\partial p_{i}}\sum_{i=1}^k \log x_{i}!+\frac{\partial {}}{\partial p_{i}}\lambda \left(1-\sum_{i=1}^{k} p_{i}\right)=0$
### $\hspace{1.73cm}= 0+\frac{\partial {}}{\partial p_{i}}\sum_{i=1}^k x_i \log p_i-0+\frac{\partial {}}{\partial p_{i}}\lambda \left(1-\sum_{i=1}^{k} p_{i}\right)$
### $\frac{\partial {}}{\partial p_{i}}\sum_{i=1}^k x_i \log p_i-\lambda\frac{\partial{}}{\partial{p_{i}}}\sum_{i=1}^{k}p_{i}=0$
### $\frac{x_{i}}{p_{i}}-\lambda=0$
## $p_{i} = \frac{x_{i}}{\lambda}$
### Since, the summation over all probabilities should sum to 1
### $\sum_{i=1}^{k}p_i = \sum_{i=1}^{k}\frac{x_{i}}{\lambda}=1$
### $\sum_{i=1}^{k}\frac{x_{i}}{\lambda}=1$
### $\frac{1}{\lambda} \sum_{i=1}^{k} x_{i} = 1$
### we know that $\hspace{0.25cm}\sum_{i=1}^{k} x_{i} = n$
### $\lambda = n$
### so $\hspace{0.5cm}p_i = \frac{x_{i}}{n}$
### The probability distribution parameters for MLE of k-sided multinomial distribution is
## $p = \left(\frac{x_1}{n},\frac{x_2}{n},\frac{x_3}{n},...,\frac{x_k}{n}\right)$